### PR TITLE
Feat(models): Implemented three models for license similarity

### PR DESCRIPTION
**Pull request by Kaushl2208**
### Description
Implementation of Logistic Regression, Multinomial Naive Bayes and Linear SVC on license dataset `licenseList.csv`.

### Files
- train.py (Training the models and saving in binary)
- test.py ( For the testing purpose)
- lr_model.pkl (Binary file for logistic regression)
- nb_model.pkl(Binary file for Multinomial Naive Bayes)
- svc_model.pkl(Binary file for Linear SVC)
- vectorizer.pkl (Binary file for storing vocabulary)

### ToDo

- Test working and accuracy  of the algorithms using `evaluator.py`
- proper integration with `atarashii.py`

CC: @hastagAB @GMishx @ag4ums

### CODE REVIEW

**Comment by Kaushl2208** [comment]
@hastagAB @GMishx , I implemented the models command into `atarashii.py` but it seems like I am missing something to update somewhere in code.

**Comment by Kaushl2208** [comment]
@GMishx @ag4ums I have run all three models on the Test files and I am attaching the screenshot of the results.
### SVC
SVC

### NB
NB

### Logistic Regression
LR

File: atarashi/atarashii.py [hunk 1]
```diff
@@ -56,6 +56,9 @@ def atarashii_runner(inputFile, processedLicense, agent_name, similarity="Cosine
   scanner = ""
   if agent_name == "wordFrequencySimilarity":
     scanner = WordFrequencySimilarity(processedLicense)
+  elif agent_name == "lr_classifier" or agent_name == "svc_classifier" or agent_name == "nb_classifier":
+    result = similarity_calc(inputFile,agent_name)
```
Comment (GMishx) [inline atarashi/atarashii.py:60]: The interface does not look consistent here. Please consider extending `AtarashiAgent` class and implement `scan` method https://github.com/fossology/atarashi/blob/23877eec31e5b76e3fcd899be337aeddd140c4d5/atarashi/agents/atarashiAgent.py#L31

You can refer other agents inside `atarashi/agents`.

File: atarashi/agents/models/test.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
+import joblib
+import os
+import pickle
```
Comment (GMishx) [inline atarashi/agents/models/test.py:3]: Please remove unused imports.

File: atarashi/agents/models/test.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import argparse
+from atarashi.libs.commentPreprocessor import CommentPreprocessor
+from sklearn.svm import LinearSVC
+from sklearn.feature_extraction.text import CountVectorizer
+from sklearn.linear_model import LogisticRegression
+from sklearn.naive_bayes import MultinomialNB
+
+def new(processed_comment,model):
```
Comment (GMishx) [inline atarashi/agents/models/test.py:11]: Please consider a better name for the method.

File: atarashi/agents/models/test.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        license_name = classifier.predict((loaded_vect.transform([processed_comment])))      
+    
+    elif model == "svc_classifier":
+        classifier = joblib.load(os.path.join(temp, 'binaryFiles/svc_model.pkl'))
+        license_name = classifier.predict((loaded_vect.transform([processed_comment])))
+
+    return license_name
+
+
+def similarity_calc(filePath, model):
```
Comment (GMishx) [inline atarashi/agents/models/test.py:33]: Please extend the `AtarashiAgent` class' method `scan` for a neat interface.

Comment (GMishx) [reply atarashi/agents/models/test.py:33]: Also, please use the `initial_match` function to get the SPDX Identifier matches https://github.com/fossology/atarashi/blob/23877eec31e5b76e3fcd899be337aeddd140c4d5/atarashi/agents/cosineSimNgram.py#L104

File: atarashi/agents/models/test.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    commentFile = CommentPreprocessor.extract(filePath)
+    with open(commentFile) as file:
+        doc = file.read()
+    processed_comment = CommentPreprocessor.preprocess(doc)
+    license_name = new(processed_comment,model)
+    
+   
+    match.append({
+      'shortname': license_name[0],
+      'sim_score': 1,
```
Comment (GMishx) [inline atarashi/agents/models/test.py:59]: Is it somehow possible to get the confidence score from the agent rather than hard coding 100% ?

File: atarashi/agents/models/train.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+from sklearn.svm import LinearSVC
+from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
+from sklearn.linear_model import LogisticRegression
+from sklearn.naive_bayes import MultinomialNB
+
+
+
+def model_train():
+
+    data = pd.read_csv("atarashi/data/licenses/licenseList.csv")
```
Comment (GMishx) [inline atarashi/agents/models/train.py:14]: Please consider using relative paths rather than absolute path like https://github.com/fossology/atarashi/blob/fc89cb1bdd678bf3c7547155bf23332f8f5ea03d/atarashi/build_deps.py#L44-L45

File: atarashi/agents/models/train.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    l_regress.fit(X_train_tfidf,y_train)
+    svc_classifier.fit(X_train_tfidf,y_train)
+    
+
+    print("Training completed")
+    print("Saving the models into data folder....")
+    joblib.dump(naive_bayes,"atarashi/agents/models/binaryFiles/nb_model.pkl")
+    joblib.dump(l_regress,"atarashi/agents/models/binaryFiles/lr_model.pkl")
+    joblib.dump(svc_classifier,"atarashi/agents/models/binaryFiles/svc_model.pkl")
+    joblib.dump(count_vect,"atarashi/agents/models/binaryFiles/vectorizer.pkl")
```
Comment (GMishx) [inline atarashi/agents/models/train.py:42]: Same relative path approach will be better here.

**Review by GMishx — CHANGES_REQUESTED** [review]

File: atarashi/agents/models/train.py (hunk 6 above)
Comment (GMishx) [inline atarashi/agents/models/train.py:14]: As a future improvement, SPDX license data can be pulled in using `atarashi.license.licenseDownloader.LicenseDownloader.download_license` and merged with main list using `atarashi.license.license_merger.license_merger`.

File: atarashi/agents/models/test.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+    '''
+    The function is designed to give output as the most similar predicted files
+    provided by the user. There are three different model approaches are designed
+    which can result into different similarities. The comments from files are
+    extracted and then the prediction is done on the basis of pre-trained
+    models in data folder.
+
+    :param filePath: Input file path to scan
+    :param model: the name of the model by which the prediction is made.
```
Comment (GMishx) [inline atarashi/agents/models/test.py:64]: I guess the parameter `model` does not exists any more.

File: atarashi/agents/models/test.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+"""
+
+import joblib
+import os
+import argparse
+from atarashi.agents.atarashiAgent import AtarashiAgent
+from atarashi.libs.initialmatch import spdx_identifer
+
+
+class Model(AtarashiAgent):
```
Comment (GMishx) [inline atarashi/agents/models/test.py:29]: Please add the documentation for the class.

File: atarashi/agents/models/test.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+  lr_classifier = "lr_classifier"
+  nb_classifier = "nb_classifier"
+  svc_classifier = "svc_classifier"
+
+  def __init__(self, licenseList, modelsLoc):
+    super().__init__(licenseList)
+    self.models_folder = os.path.abspath(modelsLoc)
+
+  def similarity_calc(self, processed_comment):
```
Comment (GMishx) [inline atarashi/agents/models/test.py:39]: Please add the documentation for the method.

File: atarashi/agents/models/train.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+import joblib
+from atarashi.libs.commentPreprocessor import CommentPreprocessor
+from sklearn.svm import LinearSVC
+from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
+from sklearn.linear_model import LogisticRegression
+from sklearn.naive_bayes import MultinomialNB
+
+
+
+def model_train():
```
Comment (GMishx) [inline atarashi/agents/models/train.py:33]: Please add the documentation for the method.

File: atarashi/atarashii.py [hunk 12]
```diff
@@ -114,13 +125,17 @@ def main():
...
   processedLicense = args.processedLicenseList
   ngram_json = args.ngram_json
+  models = args.models
 
   if processedLicense is None:
     processedLicense = defaultProcessed
   if ngram_json is None:
     ngram_json = defaultJSON
+  if ngram_json is None:
+    ngram_json = defaultModels
```
Comment (GMishx) [inline atarashi/atarashii.py:135]:
```suggestion
  if models is None:
    models = defaultModels
```

File: README.md [hunk 13]
```diff
@@ -81,6 +81,22 @@ Get the help by running `atarashi -h` or `atarashi --help`
     - With **Bigram Cosine similarity**
 
         `atarashi -a Ngram -s BigramCosineSim /path/to/file.c`
+- **Classification models**
+    - **Training** (optional)
+        
+        `python3 train.py`
```
Comment (GMishx) [inline README.md:87]: It would be nice to give complete path here.

```suggestion
        `python3 atarashi/agents/models/train.py`
```

File: README.md [hunk 14]
```diff
@@ -81,6 +81,22 @@ Get the help by running `atarashi -h` or `atarashi --help`
...
+
+        - **Logistic Regression**
+
+            `atarashi -a lr_classifier -m /path/to/models/folder/ /path/to/file.c`
+        - **Multimomial Naive Bayes**
+
+            `atarashi -a nb_classifier -m /path/to/models/folder/ /path/to/file.c`
+        - **Linear SVC** 
+
+            `atarashi -a svc_classifier -m /path/to/models/folder/ /path/to/file.c`
```
Comment (GMishx) [inline README.md:99]: Please remove the `-m` parameter as it is optional to run the agents. Instead, it can documented like **Running with custom CSVs and JSONs** point below. It would be less confusing IMO.

**Review by GMishx — CHANGES_REQUESTED** [review]
Few more changes are required. And please squash your commits.

File: README.md (hunk 14 above)
Comment (Kaushl2208) [reply README.md:99]: Yes, I also found it a little odd.

**Review by GMishx — APPROVED** [review]
Agent looks good.
Tested with `pip install .`
